openquake/hazardlib/near_fault.py

```diff
@@ -0,0 +1,476 @@
+    # same vector with PpPh,  and PpPh >= PdPh
+    # Transform to xyz coordinate
+
+    node0_xyz = get_xyz_from_ll(node0, reference)
+    node1_xyz = get_xyz_from_ll(node1, reference)
+    node2_xyz = get_xyz_from_ll(node2, reference)
+    node3_xyz = get_xyz_from_ll(node3, reference)
+    hypocenter_xyz = get_xyz_from_ll(hypocenter, reference)
+    hypocenter_xyz = np.array(hypocenter_xyz).flatten()
+
+    pp_xyz = pp
+
+    e = []
+
+    # Loop each segments on the patch to find Pd
+    segment_s = [node0_xyz, node1_xyz, node2_xyz, node3_xyz]
+    segment_e = [node1_xyz, node2_xyz, node3_xyz, node0_xyz]
+
+    # set buffering bu
+    buf = 0.0001
+    atol = 0.0001
+
+    loop = True
+    exit_flag = False
+    looptime = 0.
+    while loop:
+        x_min = np.min(np.array([node0_xyz[0], node1_xyz[0], node2_xyz[0],
+                                node3_xyz[0]])) - buf
+        x_max = np.max(np.array([node0_xyz[0], node1_xyz[0], node2_xyz[0],
+                                node3_xyz[0]])) + buf
+        y_min = np.min(np.array([node0_xyz[1], node1_xyz[1], node2_xyz[1],
+                                node3_xyz[1]])) - buf
+        y_max = np.max(np.array([node0_xyz[1], node1_xyz[1], node2_xyz[1],
+                                node3_xyz[1]])) + buf
+        n_seg = 0
+        exit_flag = False
+        for (seg_s, seg_e) in zip(segment_s, segment_e):
+
+            p_intersect, vector1, vector2, vector3, vector4 = _intersection(
+                seg_s, seg_e, pp_xyz, hypocenter_xyz)
+
+            ppph = dst.pdist([pp, hypocenter_xyz])
+            pdph = dst.pdist([p_intersect.flatten(), hypocenter_xyz])
+            n_seg = n_seg + 1
+
+            # Check that the direction of the hyp-pp and hyp-pd vectors
+            # have are the same.
+            if (np.allclose(vector1.flatten(), vector2,
+                            atol=atol, rtol=0.)):
+                if ((np.allclose(vector3.flatten(), vector4, atol=atol,
+                                 rtol=0.))):
+
+                    # Check if ppph >= pdph.
+                    if (ppph >= pdph):
+                        if (p_intersect[0] >= x_min) & (p_intersect[0] <=
+                                                        x_max):
+                            if (p_intersect[1] >= y_min) & (p_intersect[1]
+                                                            <= y_max):
+                                e = pdph
+                                pd = p_intersect
+                                exit_flag = True
+
+                                break
+            # when the pp located within the fault rupture plane, e = ppph
+            if not e:
+                if (pp_xyz[0] >= x_min) & (pp_xyz[0] <= x_max):
+                    if (pp_xyz[1] >= y_min) & (pp_xyz[1] <= y_max):
+                        pd = pp_xyz
+                        e = ppph
+                        exit_flag = True
+
+        if exit_flag:
+            break
+
+        if not e:
+
```

Comment:
I would remove this space. Can you provide a short comment explaining what you are doing here?